llvmLocalityOptimization.cpp

```diff
@@ -0,0 +1,1423 @@
+		    gf_addr_args[0] = tmpOp;
+		    newOp = CallInst::Create(fn, gf_addr_args, "", insertBefore);
+		    if (debugPassInsn) {
+			errs() << "GF Inserted : " << *newOp << "\n";
+		    }
+		}
+	    } else {
+		newOp = tmpOp;
+	    } 
+	    return newOp;
+	}
+
+	Instruction* duplicateCallInst(CallInst *oldCall, Function* newF) {
+	    Instruction *newCall;
+	    CallSite CS(oldCall);
+	    const AttributeSet &CallPAL = CS.getAttributes();
+	    SmallVector<Value*, 16> args;
+
+	    for (unsigned int i = 0; i < oldCall->getNumArgOperands(); i++) {
+		Value *op = oldCall->getArgOperand(i);
+		args.push_back(op);
+	    }
+	    
+	    if (InvokeInst *II = dyn_cast<InvokeInst>(oldCall)) {
+		newCall = InvokeInst::Create(newF, II->getNormalDest(), II->getUnwindDest(),
+					     args, "", oldCall);
+		cast<InvokeInst>(newCall)->setCallingConv(CS.getCallingConv());
+		cast<InvokeInst>(newCall)->setAttributes(CallPAL);
+	    } else {
+		newCall = CallInst::Create(newF, args, "", oldCall);
+		cast<CallInst>(newCall)->setCallingConv(CS.getCallingConv());
+		cast<CallInst>(newCall)->setAttributes(CallPAL);
+		if (cast<CallInst>(oldCall)->isTailCall())
+		    cast<CallInst>(newCall)->setTailCall();
+	    }
+	    if (MDNode *tbaa = oldCall->getMetadata(LLVMContext::MD_tbaa)) {
+		newCall->setMetadata(LLVMContext::MD_tbaa, tbaa);
+	    }
+	    if (MDNode *tbaaStruct = oldCall->getMetadata(LLVMContext::MD_tbaa_struct)) {
+		newCall->setMetadata(LLVMContext::MD_tbaa_struct, tbaaStruct);
+	    }
+	    return newCall;
+	}
+
+	bool checkNeedToWork(Instruction *insn, ValueToValueMapTy &VM) {
+	    bool needsWork = false;
+	    for(unsigned int i=0; i < insn->getNumOperands(); i++) {
+		Value *old = insn->getOperand(i);
+		ValueToValueMapTy::iterator I = VM.find(old);
+		if( I != VM.end() && I->second ) needsWork = true;
+	    }
+	    
+	    // check global
+	    if (insn->getOpcode() != Instruction::Call) {
+		for(unsigned int i=0; i < insn->getNumOperands(); i++) {
+		    Value *old = insn->getOperand(i);
+		    if( isaGlobalPointer(info, old->getType()) ) needsWork = true;
+		}
+		if( isaGlobalPointer(info, insn->getType()) ) needsWork = true;
+	    } else {
+		CallInst *call = cast<CallInst>(insn);
+		Function *F = call->getCalledFunction();
+		if (!F) return false;
+		if (isa<MemIntrinsic>(call) && isa<MemCpyInst>(call)) {
+		    Value* gDst = call->getArgOperand(0);
+		    Value* gSrc = call->getArgOperand(1);
+		    if (gDst->getType()->getPointerAddressSpace() == info->globalSpace
+			|| gSrc->getType()->getPointerAddressSpace() == info->globalSpace) {
+			needsWork = true;
+		    }			
+		} else if (F->getName().startswith(".gf.addr")) { 
+		    needsWork = true;
+		}
+	    }
+	    return needsWork;
+	}
+
+	void processInstruction(Instruction* targetInsn, SmallVector<Instruction*, 16> &deletedInsn, ValueToValueMapTy &VM, ValueTable *VN, Module &M, IGraph *G, LocalArrayInfo *LocalArraysGVN, LocalArrayInfo *LocalArraysDecl, std::vector<Value*> &NonLocals) {
```

Comment:
Could you wrap this to 80 lines? And add a comment about what this function does? (Is it transforming instructions according to the tables already created?)